### PR TITLE
Add parameter validation to SHA-512 module

### PR DESCRIPTION
**Pull request by andresag01**
## Description
Add parameter validation checks for inputs to functions from the SHA-512 module. This PR is still in development and is dependent on the API for parameter validation checks to be finalised (see [here](https://github.com/ARMmbed/mbedtls/pull/2268)).

## Status
**IN DEVELOPMENT**

## Requires Backporting
NO  

## Additional comments
The API is not finalised, so this PR is still in development and needs testing. In fact, it does not even build because the changes need to be rebased on top of the base parameter validation changes.

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

### CODE REVIEW

File: library/sha512.c [hunk 1]
```diff
@@ -88,8 +88,14 @@
 }
 #endif /* PUT_UINT64_BE */
 
+#define MBEDTLS_SHA512_VALIDATE_RET(cond)                           \
+    MBEDTLS_VALIDATE_RET( MBEDTLS_ERR_SHA512_BAD_INPUT_DATA, cond )
```
Comment (AndrzejKurek) [inline library/sha512.c]: The order of these macro arguments is wrong.

**Comment by hanno-arm** [comment]
Rebased on top of #1643, not including any rework so far.

**Comment by hanno-arm** [comment]
@k-stachowiak Please re-review.

File: include/mbedtls/sha512.h [hunk 2]
```diff
@@ -263,13 +269,14 @@ int mbedtls_sha512_ret( const unsigned char *input,
...
- *                 Must not be \c NULL.
- * \param is384    Determines which function to use:
- *                 0: Use SHA-512, or 1: Use SHA-384.
+ * \param input    The buffer holding the data. This must be a
+ *                 readable buffer of length \p ilen Bytes. It
+ *                 must not be \c NULL.
+ * \param ilen     The length of the input data \p input in Bytes.
+ * \param output   The SHA-384 or SHA-512 checksum result. This must
+ *                 be a writable buffer of length \c 64 Bytes.
+ * \param is384    Determines which function to use. This must be eiher
```
Comment (dgreen-arm) [inline include/mbedtls/sha512.h]: Typo: eiher -> either

Comment (hanno-arm) [reply include/mbedtls/sha512.h]: Thanks @dgreen-arm, I will fix this.

Please note that the question of whether we want to accept `NULL` input data is still being discussed, so this part of the PR might be reworked.

**Comment by hanno-arm** [comment]
Thanks @dgreen-arm for your review. I believe I have addressed your comments. Could you please re-review?

File: include/mbedtls/sha512.h [hunk 3]
```diff
@@ -94,23 +99,27 @@ void mbedtls_sha512_clone( mbedtls_sha512_context *dst,
...
+ *                 and have a hash operation started.
+ * \param input    The buffer holding the input data. This must
+ *                 be a readable buffer of length \p ilen Bytes.
+ * \param ilen     The length of the input data in Bytes.
  *
  * \return         \c 0 on success.
+ * \return         A negative error code on failure.
  */
 int mbedtls_sha512_update_ret( mbedtls_sha512_context *ctx,
                     const unsigned char *input,
```
Comment (sbutcher-arm) [inline include/mbedtls/sha512.h]: Alignment is out. Not a blocker.

**Comment by sbutcher-arm** [comment]
PR #1643 has been merged, therefore removing the label 'needs:preceding PR'.

Passes CI, reviewed and approved, therefore labelling as 'ready for merge'.

**Comment by sbutcher-arm** [comment]
Passes CI and approved, therefore can keep the 'ready to merge' label.

**Pull request merged**